libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs

```diff
@@ -0,0 +1,220 @@
+  -- TODO: this can be generalized when we have more examples
+  runShelleyBase $ do
+    res <- applySTS @(ConwayGOVCERT Conway) $ TRC (ConwayGovCertEnv pParams epochNo, vState, signal)
+    pure $ case res of
+      Left pfailure -> counterexample (show pfailure) $ property False
+      Right _ -> property True -- There are no actual invariants on `VState` that we can check
+                               -- so we are happy just checking that we actually do _something_ here
+  where
+    applySubst = map . substPredWithVarTest
+
+    -- General setup
+    proof        = Conway Standard
+
+    -- PParams
+    pParamsT        = Lensed (pparams proof) id
+    pParamsPs s     = applySubst s $ pParamsPreds proof
+    pParamsGen s    = genFromConstraints proof standardOrderInfo (pParamsPs s) pParamsT
+    pParamsShrink s = shrinkFromConstraints standardOrderInfo (pParamsPs s) pParamsT
+
+    -- VState
+    vStatePs s     = applySubst s $ vstatePreds proof
+    vStateGen s    = genFromConstraints proof standardOrderInfo (vStatePs s) vstateT
+    vStateShrink s = shrinkFromConstraints standardOrderInfo (vStatePs s) vstateT
+
+    validGovCert pParams _ vState cert =
+      case cert of
+        ConwayRegDRep key dep _        -> Map.notMember key (vsDReps vState) && dep == pParams ^. ppDRepDepositL
+        ConwayUnRegDRep key dep        -> fmap drepDeposit (Map.lookup key (vsDReps vState)) == Just dep
+        ConwayUpdateDRep key _         -> Map.member key (vsDReps vState)
+        ConwayAuthCommitteeHotKey _ _  -> True
+        ConwayResignCommitteeColdKey _ -> True
+    genConwayGovCert pParams _ vState =
+      oneof $ [ do key <- arbitrary `suchThat` (`Map.notMember` vsDReps vState)
+                   ConwayRegDRep key (pParams ^. ppDRepDepositL) <$> arbitrary
+              ] ++
+              [ do (k, dep) <- elements $ Map.toList (vsDReps vState)
+                   pure $ ConwayUnRegDRep k (drepDeposit dep)
+              | mempty /= vsDReps vState
+              ] ++
+              [ ConwayAuthCommitteeHotKey <$> arbitrary <*> arbitrary
+              , ConwayResignCommitteeColdKey <$> arbitrary
+              ]
+    shrinkConwayGovCert pParams epochNo vState =
+      filter (validGovCert pParams epochNo vState) . shrink
+
+------------------------------------------------------------------------
+-- Test Tree
+------------------------------------------------------------------------
+
+genUniverse :: IO (Subst (ConwayEra StandardCrypto))
+genUniverse = generate (genSubstFromConstraints (Conway Standard) standardOrderInfo (universePreds def (Conway Standard)))
+
+-- NOTE: these tests can be slow (~10 seconds) because it takes time to generate
+-- the initial universe. Reducing the number of tests will NOT help.
+tests_STS :: TestTree
+tests_STS =
+  withResource genUniverse (const $ pure ()) $ \ io_subst ->
+  testGroup "STS property tests"
+    [ testProperty "prop_GOVCERT" $ ioProperty (prop_GOVCERT <$> io_subst)
+    ]
```

Comment:
@TimSheard this is where we re-use the universe we generate. The nice thing here is that `io_subst` can be re-used in all the different properties.